### PR TITLE
Publish `http4s-client-testkit` module

### PR DESCRIPTION
**Pull request by armanbilge**
Closes https://github.com/http4s/http4s/issues/5320.

It makes sense to focus this only on clients, right? Or should it just be a generic `testkit` module that will maybe do servers too in the future?

### CODE REVIEW

File: build.sbt [hunk 1]
```diff
@@ -13,6 +13,8 @@ ThisBuild / developers += tlGitHubDev("rossabaker", "Ross A. Baker")
 ThisBuild / tlCiReleaseBranches := Seq("series/0.23")
 ThisBuild / tlSitePublishBranch := Some("series/0.23")
 
+ThisBuild / tlCiReleaseBranches += "feature/client-testkit"
```
Comment (armanbilge) [inline build.sbt:16]: Temporarily snapshotting this branch for downstream experimentation.

File: build.sbt [hunk 2]
```diff
@@ -404,6 +406,24 @@ lazy val client = libraryCrossProject("client")
...
 
+lazy val clientTestkit = libraryCrossProject("client-testkit")
+  .settings(
+    description := "Client testkit for building http4s clients",
+    startYear := Some(2014),
+    libraryDependencies ++= Seq(
+      munit.value,
+      munitCatsEffect.value,
+    ),
+    mimaPreviousArtifacts := Set.empty,
```
Comment (armanbilge) [inline build.sbt:417]: No bincompat guarantees.

File: client-testkit/shared/src/main/scala/org/http4s/client/testkit/ClientRouteTestBattery.scala [hunk 3]
```diff
@@ -205,4 +209,23 @@ abstract class ClientRouteTestBattery(name: String) extends Http4sSuite with Htt
...
+  ): Fixture[IO[A]] =
+    registerSuiteFixture(UnsafeResourceSuiteLocalDeferredFixture(name, resource))
+
+  implicit private class ParseResultSyntax[A](self: ParseResult[A]) {
+    def yolo: A = self.valueOr(e => sys.error(e.toString))
+  }
+
+  private[this] lazy val suiteFixtures = List.newBuilder[Fixture[_]]
+
+  override def munitFixtures: Seq[Fixture[_]] = suiteFixtures.result()
```
Comment (armanbilge) [inline client-testkit/shared/src/main/scala/org/http4s/client/testkit/ClientRouteTestBattery.scala:229]: Had to inline these to drop the http4s testing dependency.

File: build.sbt (hunk 2 above)
Comment (rossabaker) [reply build.sbt:417]: That's what we usually do.  We can revisit if someone is testing a backend outside `org.http4s`, but I've never heard of one.

**Comment by rossabaker** [comment]
A separate server-testkit in due time makes sense.  Server and client are themselves already separate.

**Comment by armanbilge** [comment]
👍 working in the schismed repos.

File: client-testkit/shared/src/main/scala/org/http4s/client/testkit/ClientRouteTestBattery.scala (hunk 3 above)
Comment (rossabaker) [reply client-testkit/shared/src/main/scala/org/http4s/client/testkit/ClientRouteTestBattery.scala:230]: I don't love this.  It seems like most of them are munit concerns, and either are missing from munit or us using munit inappropriately.  But I think this is sensible for now.

**Review by rossabaker — APPROVED** [review]

**Comment by armanbilge** [comment]
I'm fixing them now. The conflicts are because of the testkit package rename and you just merged my other PR deleting the various client backends :)

File: client-testkit/shared/src/main/scala/org/http4s/client/testkit/ClientRouteTestBattery.scala (hunk 3 above)
Comment (armanbilge) [reply client-testkit/shared/src/main/scala/org/http4s/client/testkit/ClientRouteTestBattery.scala:230]: Yep, agree. This weirdness gave me a heck of a time with with null pointer exceptions and initialization order issues.

**Pull request merged**